### ISSUE TITLE
HTTP 504 Gateway Timeout when uploading large documents

### ISSUE THREAD

**Issue by megies** [krischer/jane#34]
I'm trying to upload a large StationXML file with 225 stations with three channels each. After 5 minutes I'm getting a HTTP 504 "Gateway Timeout". I've tried to set high timeout values using curl (`curl --max-time 3600 ...`) and requests (`requests.put(..., timeout=3600)`), but to no avail.

My guess is that this is some internal timeout in interaction between jane and the underlying webserver? Is there a way to set a higher internal timeout? Of course, splitting up the StationXML file into smaller files is a simple workaround, but I'd rather keep the whole measurement campaign metadata in one single file..

**Comment by krischer** [comment]
You should be able to set this with the `TimeOut Directive` in Apache: http://httpd.apache.org/docs/2.2/mod/core.html#page-header

Alternatively you could just log-in and upload it with a management command - then there are no time-outs as it all happens outside of the server.

```
python manage.py upload_documents --user me stationxml data.xml
```

This is so slow as Jane, upon uploading, generates the response plots for each channel.

Also be aware that any fdsnws request at the `channel` or `response` level will be really slow as it then has to load the whole StationXML file.

**Comment by megies** [comment]
Update: This timeout already happens for a file with 55 stations with three channels each.. which is not at all overblown.

Thanks for the workarounds!

**Comment by barsch** [comment]
does this also happen if you run jane (django) locally in debug mode?

if not, there are multiple timeout settings involved depending what setup you got for your live server - so my typical setup would be django on top of a uwsgi app server on top of nginx web server - both of the latter have timeout / max upload file size settings which can be tuned

**Comment by megies** [comment]
> does this also happen if you run jane (django) locally in debug mode?

Not sure.. don't have a local instance i can play around with right now..

Right now need to do the manage.py workaround, don't have root access on the production system right now.

**Comment by barsch** [comment]
ah I just read Lions answer - running it ```python manage.py upload_documents``` works - so its not an issue with Jane - instead the app or the web server (apache here) are to blame ...

**Comment by megies** [comment]
Yes, I can confirm now, it really does work with manage.py directly.

**Comment by krischer** [comment]
We should also maybe wrap anything below this line (https://github.com/krischer/jane/blob/f486524c22d7be7c47726a5ae0b20c7cbd092df2/src/jane/documents/signals.py#L65) in a transaction.

I guess this is fairly unlikely to actually be interrupted as all the expensive stuff happens in the line before that but it cannot hurt.

@barsch Any downside to just wrapping this block in `with transaction.atomic():`?

**Comment by barsch** [comment]
personally I try to avoid bothering with transactions - so I have not to much experience with it

so RTFM: https://docs.djangoproject.com/en/1.11/topics/db/transactions/ - especially the part "Avoid catching exceptions inside atomic!" ;)
